### PR TITLE
[Core] Consistent definition of AssembleRowContribution with block builder and solver

### PR DESCRIPTION
**Pull request by loumalouomega**
Consistent definition of USE_LOCKS_IN_ASSEMBLY with the block builder and solvers

### CODE REVIEW

**Review by RiccardoRossi — COMMENTED** [review]
definitely not a FastPR

**Comment by loumalouomega** [comment]
Closing as it will duplicate another PR

**Comment by loumalouomega** [comment]
@RiccardoRossi this is a Hotfix on fact, without this the ResidualBasedEliminationBuilderAndSolver doesn't work on OMP for large problems (it is consistent with the Block B&S)

**Comment by loumalouomega** [comment]
In fact is more complex than that, I will implement the equivalent to AssembleRowContribution for the EliminationBuilderAndSolver

**Comment by loumalouomega** [comment]
Now the AssembleRowContribution equivalent for the EliminationBuilderAndSolver is implemented

**Comment by loumalouomega** [comment]
Still some problems in the assembling of large problems with this B&S on OMP, @RiccardoRossi do I apply the corrections here?

**Comment by RiccardoRossi** [comment]
the residualbased elimination builder and solver was working perfectly fine in openmp...

**Comment by loumalouomega** [comment]
The problem of the falling ball without mpc (so just using the regular elimination builder and solver) doesn't work in OMP(I mean in the master) . In fact until last week we didn't have any test using the elimination builder and solver. It is completely normal that in certain point stopped working.

**Comment by RiccardoRossi** [comment]
so you are tellign that the elimination builder and solver in master does not work without openmp?

**Comment by loumalouomega** [comment]
Apparently not

**Comment by RiccardoRossi** [comment]
...can u give me a simple example (not even a test, simply an example) which fails in the master?

**Comment by loumalouomega** [comment]
I rerun my example, now looks like it work perfectly fine. In any case this PR adds the equivalent to AssembleRowContribution that is used on #3249 

And I already created a dummy element on the cpp test #3236

**Comment by loumalouomega** [comment]
After merging #3236 I think we can proceed with this one

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h [hunk 1]
```diff
@@ -16,17 +16,12 @@
 
 /* System includes */
 #include <set>
+#include <unordered_set>
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: note that to get the behaviour you had before, you need to actually define USE_LOCKS_IN_ASSEMBLY

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h [hunk 2]
```diff
@@ -235,7 +235,7 @@ class ResidualBasedEliminationBuilderAndSolverComponentwise
                     EquationId[i] = geom[i].GetDof(rVar,pos).EquationId();
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h]: also, to get the same behaviour we need to consistently define USE_LOCKS_IN_ASSEMBLY

**Review by RiccardoRossi — COMMENTED** [review]
@loumalouomega you are changing the behaviour by not defning the USE_LOCKS_ON_ASSEMBLY

is this the reason for which you marked this as BehaviourChange?

**Comment by jcotela** [comment]
Just a comment: the `_OPENMP` flag is the official way to know if the compilation has openmp support. If you rename those macros, please make sure that the "locking" behavior can only be used if we have openmp, otherwise it will be a compilation error when we don't.

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: As it is in the block_builder_and_solver:

https://github.com/KratosMultiphysics/Kratos/blob/503f7b8073816a65696819af7c5f4aa1ef450dea/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L27

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h]: As I said, I am being consistent with the block_builder_and_solver. Maybe that should be changed there too

**Comment by loumalouomega** [comment]
@jcotela I am being consistent with the block_builder_and_solver:

https://github.com/KratosMultiphysics/Kratos/blob/503f7b8073816a65696819af7c5f4aa1ef450dea/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L27

Maybe we should check the implementation there too

**Comment by jcotela** [comment]
@loumalouomega @RiccardoRossi is there a mechanism to set the `USE_LOCKS_IN_ASSEMBLY` somewhere in kratos or is this something that is only enabled if the flag is passed through cmake? The only thing I am saying is that, if it is automated, it should be conditional on having `_OPENMP` defined too.

**Comment by loumalouomega** [comment]
The thing is that @RiccardoRossi  wants to get rid of the lock, and basically with the current implementation can be removed safely. I just put it there for consistency until total agreement

**Comment by RiccardoRossi** [comment]
@jcotela i see your point, but there is no check that USE_LOCKS is used consistently with openmp.
maybe we should put something like

     #ifdef USE_LOCKS
            #ifnded _OPENMP
                   compilation error with meaningful message

Having said this, i do think that we could remove the use of locks since from recent tests atomics are fast.

There is no objection from my side (**actually i am convinced it is very positive!**) in changing the behaviour of the EliminationBuilderAndSolver HOWEVER this suddently became a very important change since it affects anyone who uses the EliminationBuilderAndSolver

We should 
1 - verify running with very many threads that that there is no race condition (this could be achieved with the small test you prepared and merged. @loumalouomega can u say which is the test exactly? @jcotela @pooyan-dadvand you are the ones with the higher thread count...could u try that on your machine?)
2 - prepare a large test and verify build times BOTH IN WINDOWS and in LINUX (with the elimination builder and solver)

**Comment by loumalouomega** [comment]
Here https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/strategies/builder_and_solvers/test_builder_and_solver.cpp

**Comment by loumalouomega** [comment]
@RiccardoRossi which is the conclusion here?

**Comment by pooyan-dadvand** [comment]
OK, I will try to run it on my machine but I could not get what is the new mechanism

**Comment by philbucher** [comment]
do you still need this PR?
It is not changing anything ...

**Comment by loumalouomega** [comment]
I think it was already merged with a different PR

**Pull request closed without merging**